javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,486 @@
+
+  }
+
+  private class DecodingInputStream(in: InputStream, table: Array[Int], ignoreInvalid: Boolean)
+    extends FilterInputStream(in) {
+
+    final val DecodeState18 = 0
+    final val DecodeState12 = 1
+    final val DecodeState14 = 2
+    final val DecodeState16 = 3
+
+    private val oneBuf = new Array[Byte](1)
+
+    private var closed = false
+    private var eof = false
+    private var out = 0
+    private var shift = DecodeState18
+
+    override def read(): Int = if (read(oneBuf, 0, 1) == -1) -1 else oneBuf(0) & 0xff
+
+    override def read(b: Array[Byte], off: Int, len: Int): Int = {
+      // scalastyle:off return
+      if (closed) throw new IOException("Stream is closed")
+      if (off < 0 || len < 0 || len > b.length - off) throw new IndexOutOfBoundsException
+      if (eof) return -1
+      var written = 0
+      // scalastyle:on return
+
+      @inline
+      def writeValue(i: Int): Int = {
+        // max value means we're writing remaining bytes after eof, no table lookup
+        if (i == Integer.MAX_VALUE) 0 else table(i) match {
+          case -1 =>
+            if (ignoreInvalid) 0
+            else throw new IOException("Illegal base64 character " + Integer.toString(i, 16))
+
+          case v =>
+            shift match {
+              case DecodeState18 =>
+                out |= v << 18
+                shift = DecodeState12
+                0
+
+              case DecodeState12 =>
+                out |= v << 12
+                b(off+written) = (out >> 16).toByte
+                out <<= 8
+                shift = DecodeState14
+                1
+
+              case DecodeState14 =>
+                out |= v << 14
+                b(off+written) = (out >> 16).toByte
+                out <<= 8
+                shift = DecodeState16
+                1
+
+              case DecodeState16 =>
+                out |= v << 16
+                b(off+written) = (out >> 16).toByte
+                out = 0
+                shift = DecodeState18
+                1
+            }
+        }
+      }
+
+      @inline
+      def endOfFile(): Int = {
+        eof = true
+        shift match {
+          case DecodeState18 => 0 // nothing
+          case DecodeState12 => throw new IOException("Base64 stream has one un-decoded dangling byte.")
+          case _             => writeValue(Integer.MAX_VALUE)
+        }
+      }
+
+      @inline
+      def padding(): Int = {
+        eof = true
+        val s = shift
+        if(s == DecodeState18 || s == DecodeState12 || (s == DecodeState14 && in.read() != '=' && !ignoreInvalid))
+          throw new IOException ("Illegal base64 ending sequence")
+        else
+          writeValue(Integer.MAX_VALUE)
+      }
+
+      def iterate(): Unit = {
+        // scalastyle:off return
+        while (written < len)
+          in.read() match {
+            case -1 =>
+              written += endOfFile()
+              return
```

Comment:
You can get rid of this `return` by turning `iterate()` into a tail-recursive method:
```scala
@tailrec
def iterate(): Unit = {
  if (written < len) {
    in.read() match {
      case -1 =>
        written += endOfFile()
      case '=' =>
        written += padding()
        iterate()
      case int =>
        written += writeValue(int)
        iterate()
    }
  }
}
```